### PR TITLE
Fix incorrect format string arguments

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Handles issues like wide-char strings being passed as `%s`, narrow-char strings being passed as `%S`, 64-bit integers being passed as `%d`, etc. It's likely we'll need to make a second pass at these in a few weeks so that we can cover other configurations, but this is an initial list of matches from my dev box.

Fixes violation of CodeQL rule __cpp/wrong-type-format-argument__ (https://lgtm.com/rules/2160310550/).

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
Clearly that worked better on my dev box than in CI.

**Comment by GrabYourPitchforks** [comment]
This is blocked by https://github.com/dotnet/runtime/issues/67470. I can work around it for now by creating a custom header definition, but this isn't a workable long-term solution since I'll eventually need to import other C++ headers like *\<type_traits\>* while working on other CodeQL rules.

**Comment by GrabYourPitchforks** [comment]
@dotnet/macios Do you have thoughts on why the OSX build might be failing? When importing `<cstdint>`, this is the dump I see ([full log here](https://dev.azure.com/dnceng/public/_build/results?buildId=1695961&view=logs&j=f2a23cee-3431-50cb-c21a-1d5929055501&t=8aeffa89-f24d-58b0-d0c4-49f806ab2e85&l=2127)).

```txt
/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/cstdint:162:8: error: no member named 'int_least8_t' in the global namespace
using::int_least8_t;
     ~~^
/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/cstdint:163:8: error: no member named 'int_least16_t' in the global namespace
using::int_least16_t;
     ~~^
/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/cstdint:164:8: error: no member named 'int_least32_t' in the global namespace
using::int_least32_t;
     ~~^
/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/cstdint:165:8: error: no member named 'int_least64_t' in the global namespace
using::int_least64_t;
     ~~^
...
```

As far as I'm aware there's nothing special going on here; all I'm doing is referencing a well-known C++ header file.

The Linux builds are failing for a similar reason.

**Comment by jkotas** [comment]
> Do you have thoughts on why the OSX build might be failing?

You cannot use standard C++ headers under src\coreclr. Most of the code under src/corclr can only use CoreCLR PAL (ie headers under src/coreclr/pal/inc today).

This is what needs to happen to remove that restriction: https://github.com/dotnet/runtime/issues/37310#issuecomment-637618212

**Comment by jkotas** [comment]
cc @AaronRobinsonMSFT

File: src/coreclr/tools/metainfo/mdinfo.cpp
Comment (jkotas) [inline src/coreclr/tools/metainfo/mdinfo.cpp:2043]: You can just use the Windows-specific `%#I64x` here (this code runs on the Windows emulator today, so the Windows-specific format string is fine) and then you won't need the `cinttypes` header that is causing the build breaks on Unix.

File: src/coreclr/tools/metainfo/mdinfo.cpp
Comment (AaronRobinsonMSFT) [inline src/coreclr/tools/metainfo/mdinfo.cpp:364]: This should be `%zu` since `size_t` is unsigned.

File: src/native/corehost/bundle/extractor.cpp
Comment (AaronRobinsonMSFT) [inline src/native/corehost/bundle/extractor.cpp:182]: `%zd` => `%zu`

File: src/native/corehost/deps_format.cpp
Comment (AaronRobinsonMSFT) [inline src/native/corehost/deps_format.cpp:101]: `%zd` => `%zu`

File: src/coreclr/tools/metainfo/mdinfo.cpp
Comment (am11) [reply src/coreclr/tools/metainfo/mdinfo.cpp:364]: If zd was changed based on the note that follows C4477, then it is probably the same think I noticed while working on a recent PR. So I decided to report it upstream: https://developercommunity.visualstudio.com/t/Note-that-follows-warning-C4477-is-inacc/1691095?space=62&q=inaccurate. The VCR team thinks the suggestion is correct and  clang/gcc are wrong about error. IMO, cl.exe is wrong; if we want printer to use signed data we can `ssize_t` cast the variable and use `%zd` but the format specifier should only correspond to the data type in use and produce suggestion accordingly. I tried. 🤷

However, using `%zu` make both worlds happy. cl.exe only _suggests_ %zd, but it is fine with %zu (i.e. C4477 goes away 🙏).

**Comment by jkotas** [comment]
The tests are failing with:
```
      Assert failure(PID 18953 [0x00004a09], Thread: 18968 [0x4a18]): Consistency check failed: _vsnwprintf_s failed. Potential globalization bug.FAILED: 0
          File: /__w/1/s/src/coreclr/utilcode/sstring.cpp Line: 2114
          Image: /datadisks/disk1/work/AA1009BA/p/corerun
```

File: src/coreclr/vm/clrex.cpp
Comment (jkotas) [inline src/coreclr/vm/clrex.cpp:1141]:
```suggestion
    result.Printf(W("%S (message resource %s)"),
```
This may fix the test failure.

Comment (GrabYourPitchforks) [reply src/coreclr/vm/clrex.cpp:1141]: The stack trace shows that the failure is [this line](https://github.com/dotnet/runtime/pull/67464/files#diff-50e0b3b5e4cd857b50daedc613103e1d07875866995e464604945c758382960cR523):

```diff
- strArgument.Printf(W("0x%x"), pInstruction->uArg);
+ strArgument.Printf(W("0x%zx"), pInstruction->uArg);
```

But since `ILInstruction.uArg` is typed as `UINT_PTR`, it seems like the new line should be correct?

I'm pretty sure that in context, this value always fits into 32 bits anyway, so maybe the answer is simply to cast back to _uint32_ and leave the original format string as-is.

Comment (jkotas) [reply src/coreclr/vm/clrex.cpp:1141]: The assert says that `_vsnwprintf_s` failed with unexpected error. Notice that all failures with this error are on non-Windows. I think it means that `_vsnwprintf_s` implementation in the Unix PAL is missing handling of `zx` format string.

Comment (GrabYourPitchforks) [reply src/coreclr/vm/clrex.cpp:1141]: Would it be ok for me to tweak https://github.com/dotnet/runtime/blob/main/src/coreclr/pal/src/safecrt/output.inl to recognize this pattern as part of this PR?

**Comment by GrabYourPitchforks** [comment]
@jkotas @AaronRobinsonMSFT Any final thoughts? The tip that the coreclr subdir uses a special copy of the CRT functions really helped unblock me on this. 👍

**Review by jkotas — APPROVED** [review]
LGTM

**Pull request merged**
